### PR TITLE
 Prevent crashes for opencv exception and std::exception

### LINKED ISSUE #8663: terminate called after throwing an instance of 'std::out_of_range’
## Description
(Brief description of the problem in no more than 2 sentences.)

## Environment info (Required)
Python 2.7.13 Numpy 1.13.1 MXNet 0.12rc0 

## Steps to reproduce
```python
import argparse

import mxnet as mx
import numpy as np
from mxnet import metric
from mxnet.initializer import Uniform

import mxdleio

parser = argparse.ArgumentParser(description="Train RNN on NC data",
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('--batch-size', type=int, default=32,
                    help='the batch size.')


def custom_fit(self, train_data, eval_metric='acc', kvstore='local', optimizer='sgd',
               optimizer_params=(('learning_rate', 0.01),), initializer=Uniform(0.01),
               arg_params=None, aux_params=None, allow_missing=False, force_rebind=False,
               force_init=False, begin_epoch=0, num_epoch=None, ):
    self.bind(data_shapes=train_data.provide_data, label_shapes=train_data.provide_label,
              for_training=True, force_rebind=force_rebind)

    self.init_params(initializer=initializer, arg_params=arg_params, aux_params=aux_params,
                     allow_missing=allow_missing, force_init=force_init)
    self.init_optimizer(kvstore=kvstore, optimizer=optimizer,
                        optimizer_params=optimizer_params)

    if not isinstance(eval_metric, metric.EvalMetric):
        eval_metric = metric.create(eval_metric)

    # training loop
    for epoch in range(begin_epoch, num_epoch):
        eval_metric.reset()
        nbatch = 0
        data_iter = iter(train_data)
        end_of_batch = False
        next_data_batch = next(data_iter)
        while not end_of_batch:
            data_batch = next_data_batch

            logging.info(data_batch.data[0].shape)
            logging.info(data_batch.provide_data)
            logging.info(data_batch.label[0].shape)
            logging.info(data_batch.provide_label)

            assert self.binded and self.params_initialized
            self.switch_bucket(data_batch.bucket_key, data_batch.provide_data,
                               data_batch.provide_label)
            raise NotImplementedError
            self._curr_module.forward(data_batch, is_train=True)
            self.backward()

            self.update()
            try:
                # pre fetch next batch
                next_data_batch = next(data_iter)
                self.prepare(next_data_batch)
            except StopIteration:
                end_of_batch = True

            self.update_metric(eval_metric, data_batch.label)

            nbatch += 1
        # sync aux params across devices
        arg_params, aux_params = self.get_params()
        self.set_params(arg_params, aux_params)

        # end of 1 epoch, reset the data-iter for another epoch
        train_data.reset()


if __name__ == '__main__':
    import logging
    import os

    head = '%(asctime)-15s %(message)s'
    logging.basicConfig(level=logging.DEBUG, format=head)
    np.set_printoptions(precision=4, suppress=True, linewidth=150, threshold=np.inf)
    logging.info("pid=%s", os.getpid())

    args = parser.parse_args()
    logging.info("%s", args)

    buckets = [2000, 3500]
    data_train = mxdleio.NCBucketIter(TRAIN_PATH, args.batch_size, neps=0, buckets=buckets, shuffle=True,
                                      )
    stack = mx.rnn.SequentialRNNCell()
    stack.add(mx.rnn.FusedRNNCell(512, num_layers=1, mode='lstm', prefix='lstm_l3', bidirectional=True))


    def sym_gen(seq_len):
        data = mx.sym.Variable('data')
        label = mx.sym.Variable('softmax_label')
        logging.info("seq_len: %s", seq_len)
        stack.reset()
        outputs, states = stack.unroll(seq_len, inputs=data, merge_outputs=True)
        pred = mx.sym.Reshape(outputs, shape=(-1, 1024))
        pred = mx.sym.FullyConnected(data=pred, num_hidden=42)
        l = mx.sym.Reshape(data=label, shape=(-1,))
        l = mx.sym.Cast(data=l, dtype='int32')

        sm = mx.sym.WarpCTC(data=pred, label=l, label_length=seq_len, input_length=seq_len)
        return sm, ('data',), ('softmax_label',)


    contexts = mx.gpu(0)
    model = mx.mod.BucketingModule(
        sym_gen=sym_gen,
        default_bucket_key=data_train.default_bucket_key,
        context=contexts)

    custom_fit(model,
               train_data=data_train,
               num_epoch=1,
               )
```

result
```

2017-11-15 14:56:20,641 pid=17540
2017-11-15 14:56:20,643 Namespace(batch_size=10)
2017-11-15 14:56:20,696 Data Length: 2972
2017-11-15 14:56:20,696 Total Samples: 281241
2017-11-15 14:56:20,696 Input Size: 80
2017-11-15 14:56:20,697 seq_len: 3500
lib/python2.7/site-packages/mxnet/rnn/rnn_cell.py:675: UserWarning: NTC layout detected. Consider using TNC for FusedRNNCell for faster speed
  warnings.warn("NTC layout detected. Consider using "
2017-11-15 14:56:20,699 seq_len: 3500
2017-11-15 14:56:25,232 (10L, 2000L, 80L)
2017-11-15 14:56:25,234 [DataDesc[data,(10L, 2000L, 80L),<type 'numpy.float32'>,NTC]]
2017-11-15 14:56:25,234 (10L, 2000L, 1L)
2017-11-15 14:56:25,234 [DataDesc[softmax_label,(10L, 2000L, 1L),<type 'numpy.float32'>,NTC]]
2017-11-15 14:56:25,234 seq_len: 2000
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
script: line 17: 17540 Aborted                 python minimal_sym_task2.py --batch-size 10


```

### LINKED ISSUE #9475: OpenCV Error: Assertion failed (dst.cols < SHRT_MAX && dst.rows < SHRT_MAX && src.cols < SHRT_MAX && src.rows < SHRT_MAX) in remap, file /home/travis/build/dmlc/mxnet-distro/deps/opencv-3.3.0/modules/imgproc/src/imgwarp.cpp, line 4944 terminate called after throwing an instance of 'cv::Exception'
When I use augmentation of aspect, random scale, rotate, mxnet report this error:
```
OpenCV Error: Assertion failed (dst.cols < SHRT_MAX && dst.rows < SHRT_MAX && src.cols < SHRT_MAX && src.rows < SHRT_MAX) in remap, file /home/travis/build/dmlc/mxnet-distro/deps/opencv-3.3.0/modules/imgproc/src/imgwarp.cpp, line 4944 terminate called after throwing an instance of 'cv::Exception'
```

I install the mxnet by
```
pip install  mxnet-cu80==1.0.1b20180105
```

I think mxnet should catch opencv Exceptions rather than terminating the process.

### LINKED ISSUE #10389: Report clear errors when opencv::imdecode fails.
Currently when opencv function fails cryptic opencv error is raised. We should catch it and report which file failed.

see https://github.com/apache/incubator-mxnet/blob/121adb6660897d7cef223caba80a43412ab5c9e0/src/io/image_io.cc#L161

### LINKED ISSUE #13217: Handle OpenCV exceptions instead of aborting
(This is intended as an umbrella issue to follow up my ongoing proposal on the dev mailing list & work on handling OpenCV exceptions)

## Description

I am working on migrating a classification product currently using Caffe to
MXNet. Along the way I'm encountering some issues loading and augmenting
the images dataset.

Basically it seems my dataset contains some technically invalid images.
When loading them using mx.io.ImageRecordIter (from a Python script), they
get passed eventually to the OpenCV library which will throw a C++
exception. MXNet currently doesn't capture those, resulting in my script
aborting with a not very clear error message.

## Error Message:
"
terminate called after throwing an instance of 'cv::Exception'

  what():  OpenCV(3.4.3)
/home/lgo/dev/opencv-3.4.3/modules/imgproc/src/resize.cpp:4044: error:
(-215:Assertion failed) !ssize.empty() in function 'resize'

Aborted (core dumped)
"

## Environment info (Required)

Linux (Ubuntu 18.04), OpenCV 3.4.3, mxmet master branch, Python 3.6.6

## Build info (Required if built from source)

Compiler (gcc/clang/mingw/visual studio):
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0

## Steps to reproduce
It's not clear exactly which image causes this particular exception. As the process aborts due to the unhandled exception, there is no easy way to capture good information about the process status - in this case the image name.

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##
fixes #8663 , fixes #9475, fixes #10389, fixes #13217

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)

### Changes ###
- [x] Added support for std::exception handling to prevent crash
- [x] Added test

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Bug, OpenCV, Exception Handling, pr-awaiting-review]

**Comment by anirudh2290** [comment]
This is dependent on waitall PR: https://github.com/apache/incubator-mxnet/pull/14397. Please review that one first.

File: src/resource.cc [hunk 1]
```diff
@@ -190,12 +190,14 @@ class ResourceManagerImpl : public ResourceManager {
     cpu_rand_->Seed(seed);
     cpu_parallel_rand_->Seed(seed);
 #if MXNET_USE_CUDA
-    gpu_rand_.Get(ctx.dev_id, [ctx, seed, this]() {
-      return new ResourceRandom<gpu>(ctx, seed);
-    })->Seed(seed);
-    gpu_parallel_rand_.Get(ctx.dev_id, [ctx, seed, this]() {
-      return new ResourceParallelRandom<gpu>(ctx, gpu_native_rand_copy_, seed);
-    })->Seed(seed);
+    if (ctx != Context::CPU()) {
```
Comment (wkcn) [inline src/resource.cc]: ctx.dev_type == kGPU may be better.

**Comment by piyushghai** [comment]
@anirudh2290 Can you rebase with the master branch and resolve merge conflicts ?

**Comment by anirudh2290** [comment]
@piyushghai done ! this is ready for review.

File: include/mxnet/c_api_error.h [hunk 2]
```diff
@@ -31,9 +31,35 @@
...
  */
-#define MX_API_BEGIN() try { on_enter_api(__FUNCTION__);
-#define MX_API_END() } catch(dmlc::Error &_except_) { on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0;  // NOLINT(*)
-#define MX_API_END_HANDLE_ERROR(Finalize) } catch(dmlc::Error &_except_) { Finalize; on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0; // NOLINT(*)
+#define MX_API_BEGIN()                                                         \
+  try {                                                                        \
+    on_enter_api(__FUNCTION__);
+#define MX_API_END()                                                           \
+  }                                                                            \
+  catch (dmlc::Error &_except_) {                                              \
```
Comment (larroy) [inline include/mxnet/c_api_error.h]: Why do we need both? isn't dmlc::Error runtime_error? single catch should work.

Comment (anirudh2290) [reply include/mxnet/c_api_error.h]: Good catch! I had kept this since i was planning on different error codes for frontend and this would differentiate between mxnet and non mxnet code. But since we are returning -1 for all currently I can remove it for now.

**Comment by anirudh2290** [comment]
@apeforest can you take a look too.

File: include/mxnet/c_api_error.h [hunk 3]
```diff
@@ -31,9 +31,26 @@
  * and finishes with API_END() or API_END_HANDLE_ERROR()
  * The finally clause contains procedure to cleanup states when an error happens.
  */
-#define MX_API_BEGIN() try { on_enter_api(__FUNCTION__);
-#define MX_API_END() } catch(dmlc::Error &_except_) { on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0;  // NOLINT(*)
-#define MX_API_END_HANDLE_ERROR(Finalize) } catch(dmlc::Error &_except_) { Finalize; on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0; // NOLINT(*)
+#define MX_API_BEGIN()                                                         \
+  try {                                                                        \
+    on_enter_api(__FUNCTION__);
+#define MX_API_END()                                                           \
```
Comment (larroy) [inline include/mxnet/c_api_error.h]: is it possible to have the defines inside the scopes a bit more ordered?

File: src/engine/threaded_engine.h [hunk 4]
```diff
@@ -384,6 +384,10 @@ class ThreadedEngine : public Engine {
           threaded_opr->opr_exception =
               std::make_shared<std::exception_ptr>(std::current_exception());
           callback();
+        } catch (std::exception& e) {
```
Comment (yuxihu) [inline src/engine/threaded_engine.h]: Can we just replace the `catch (dmlc::Error& e)` on L383 with `catch (const std::exception& e)` here?

File: include/mxnet/c_api_error.h [hunk 5]
```diff
@@ -31,9 +31,26 @@
...
+  }                                                                            \
+  catch (std::exception &_except_) {                                           \
+    on_exit_api();                                                             \
+    return MXAPIHandleException(_except_);                                     \
+  }                                                                            \
+  on_exit_api();                                                               \
+  return 0; // NOLINT(*)
+#define MX_API_END_HANDLE_ERROR(Finalize)                                      \
+  }                                                                            \
+  catch (std::exception &_except_) {                                           \
```
Comment (yuxihu) [inline include/mxnet/c_api_error.h]: const?

File: include/mxnet/c_api_error.h [hunk 6]
```diff
@@ -31,9 +31,26 @@
...
  */
-#define MX_API_BEGIN() try { on_enter_api(__FUNCTION__);
-#define MX_API_END() } catch(dmlc::Error &_except_) { on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0;  // NOLINT(*)
-#define MX_API_END_HANDLE_ERROR(Finalize) } catch(dmlc::Error &_except_) { Finalize; on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0; // NOLINT(*)
+#define MX_API_BEGIN()                                                         \
+  try {                                                                        \
+    on_enter_api(__FUNCTION__);
+#define MX_API_END()                                                           \
+  }                                                                            \
+  catch (std::exception &_except_) {                                           \
```
Comment (yuxihu) [inline include/mxnet/c_api_error.h]: const？

File: include/mxnet/c_api_error.h (hunk 3 above)
Comment (anirudh2290) [reply include/mxnet/c_api_error.h]: can you please elaborate on what you mean by that

**Comment by anirudh2290** [comment]
@larroy and @yuxihu . Thanks a lot for your review. I have addressed/responded to your comments.

**Comment by larroy** [comment]
I approved already.

File: include/mxnet/c_api_error.h (hunk 3 above)
Comment (larroy) [reply include/mxnet/c_api_error.h]: Never mind, I think it's a separate discussion from this PR.

**Comment by anirudhacharya** [comment]
@mxnet-label-bot update [pr-awaiting-merge]

**Comment by wkcn** [comment]
Merged. Thank you!
